Skip emails for the Jetpack org owner user

pandah3 opened the pull request:
## Description

This PR adds logic to skip email notifications for the [Jetpack organization owner user](https://github.com/Automattic/vip-go-mu-plugins/blob/57c86e4ec4d088796d06a9e865f75eb12402675d/vip-jetpack/connection-pilot/class-jetpack-connection-attendant.php#L173). Users don't need to get the privileged user email notification because this is an automated user for Jetpack connection setup activities.

## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [ ] This change works and has been tested locally or in Codespaces (or has an appropriate fallback).
- [x ] This change has relevant unit tests (if applicable).
- [ ] This change has relevant documentation additions / updates (if applicable).
- [ ] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test

1. Check out PR.
1. Test that emails send as expected for regular users:

```
vip dev-env exec -- wp user create testadminuser joanne.lee+adminadmin3@automattic.com --user_pass=password123345

vip dev-env exec -- wp user set-role testadminuser administrator --url=https://vip-security-boost.vipdev.lndo.site/
```

3. Make sure you get an email in Mailpit: http://vip-security-boost-mailpit.vipdev.lndo.site/
4. Create the Jetpack org owner user:

```
vip dev-env exec -- wp eval 'var_export( wp_insert_user([
  "user_login"=>"wpvip-jetpack-connection-owner",
  "user_email"=>"vip-jetpack-owner+0@wpvip.com",
  "display_name"=>"WPVIP Jetpack Connection Owner",
  "role"=>"administrator",
  "user_pass"=>"password"
]) );'
```

_P.S - The usual `wp user create wpvip-jetpack-connection-owner` won't work here bc you'll get an error: `Error: Usernames can only contain lowercase letters (a-z) and numbers.`_

5. Make sure you don't get a new email in Mailpit
6. Upgrade user to Super Admin: `vip dev-env exec -- wp super-admin add wpvip-jetpack-connection-owner`
7. Make sure you don't get a new email
8. Also test with the Jetpack user but an email address that doesn't use an A8C-owned domain:

```
// Remove old user from DB (WP-CLI commands aren't working for some reason)
vip dev-env exec -- wp cache flush

// Re-add user but with a test domain:
vip dev-env exec -- wp eval 'var_export( wp_insert_user([
  "user_login"=>"wpvip-jetpack-connection-owner",
  "user_email"=>"vip-jetpack-owner+0@test.com",
  "display_name"=>"WPVIP Jetpack Connection Owner",
  "role"=>"administrator",
  "user_pass"=>"password"
]) );'
10. Make sure you _do_ get an email
```

**Side note:** I tried testing with the Jetpack connection flow itself to mimic the behavior, but we can't do that on dev-env:

```
// using shell bc adding this to `/client-mu-plugins/force-jetpack-org-owner.php` wasn't calling the filter for some reason
vip dev-env shell
mkdir -p /wp/wp-content/mu-plugins
cat > /wp/wp-content/mu-plugins/force-jetpack-org-owner.php <<'PHP'
<?php
add_filter('vip_jetpack_connection_pilot_use_org_owner', '__return_true');
error_log('MU loaded: force-jetpack-org-owner');
PHP

// Connect JP
vip dev-env exec -- wp jetpack-start connect --url=https://vip-security-boost.vipdev.lndo.site/
Warning: ❌ Could not connect Jetpack. Error (jp-cxn-pilot-invalid-environment): This is not a valid WPVIP environment.
```

pandah3 committed:
Skip email notification for the Jetpack connection owner

pandah3 committed:
Add tests to test email skipping for Jetpack owner

pandah3 committed:
Fix lint errors

pandah3 commented on modules/notify-privileged-activity/class-notify-privileged-activity.php:
```diff
@@ -108,6 +114,15 @@ public static function notify_user_granted_super_admin( $user_id ) {
+			'Skipping notification for Jetpack connection owner: ' . $user->user_login
+		);
+		return;
+	}
```
Fixed in [5d7338c](https://github.com/Automattic/vip-security-boost/pull/97/commits/5d7338cc4f7ec61f6a0bd2472f44acab93d2e96c)

andrea-sdl commented on modules/notify-privileged-activity/class-notify-privileged-activity.php:
```diff
@@ -8,6 +8,12 @@
+
+	private static function is_jetpack_connection_owner( $user ): bool {
+		return ( $user instanceof \WP_User )
+			&& 'wpvip-jetpack-connection-owner' === strtolower( $user->user_login );
```
nitpick: @pandah3 we bring the `wpvip-jetpack-connection-owner` into a constant to make referencing it easier? I was also wondering if we should also check that the email ends with `@wpvip.com` to make sure that, if a site doesn't have this user and someone adds it, we'll send the email if the email address doesn't match one owned by VIP (we could also check for `@automattic.com` too if we want)

pandah3 replied:
Addressed both in [df6f7f6](https://github.com/Automattic/vip-security-boost/pull/97/commits/df6f7f6a6280224c47e1b6b5ab3a6c41ba51720a)!

andrea-sdl reviewed (approved):
Works as expected, thank you for the testing instructions! I left a comment but I think it's already good to go.

pandah3 committed:
Check for valid a8c/vip emails for the Jetpack org owner to ensure we're not skipping emails when we shouldn't

pandah3 committed:
Load class in plugin bootstrap

pandah3 committed:
Ignore lint error for having an inline comment

pandah3 committed:
Fix failing tests

brunobasto commented on modules/notify-privileged-activity/class-notify-privileged-activity.php:
```diff
@@ -2,12 +2,27 @@
+	 *  - Email domain is A8C/VIP owned
+	 */
+	private static function is_jetpack_connection_owner( $user ): bool {
+		return ( $user instanceof \WP_User )
```
nit: I think expanding this to use if/else blocks would make it easier to understand :)

pandah3 replied:
Updated in [26d837a](https://github.com/Automattic/vip-security-boost/pull/97/commits/26d837a0d783260feec08b3ecf33288be082053c)!

brunobasto commented on utils/class-email-utils.php:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	/** True if the email is on an Automattic/VIP-owned domain (filterable) */
+	public static function is_a8c_owned( $email ): bool {
```
Maybe we can use this existing utility from mu-plugins instead? https://github.com/Automattic/vip-go-mu-plugins/blob/develop/vip-support/class-vip-support-user.php#L760

pandah3 replied:
Ooh nice!! Updated to use that instead in [1463705](https://github.com/Automattic/vip-security-boost/pull/97/commits/14637051d44547373ca3b14c7e50c8945b81eaca). Thank you!!

pandah3 committed:
Refactor function to make it easier to read

pandah3 committed:
Use util from mu-plugins to check for A8C-owned email domains

pandah3 committed:
Fix tests:
- Add mock class to test bootstrap so it loads during tests
- Add missing mock for is_verified_automattician
Check to make sure the class and method from mu-plugins exist

pandah3 committed:
Fix phpstan error

brunobasto reviewed (approved):
Looks good to me! 🚢
